### PR TITLE
Digitizer switch usages support

### PR DESCRIPTION
**Pull request by avg-I**
The code could probably be better, but at least it works for me.

### CODE REVIEW

File: hconf.c [hunk 1]
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
...
+}
+
+static int
+hconf_switch_handler(SYSCTL_HANDLER_ARGS)
+{
+	struct hconf_softc *sc = arg1;
+	u_int value;
+	int error;
+
+	error = hconf_get_switch(sc, arg2, &value);
```
Comment (wulf7) [inline hconf.c]: Microsoft does not require configuration page feature reports to be readable:
https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/touchpad-required-hid-top-level-collections
This means that old values should be taken from soft context rather then from hardware. hconf_get_switch() can be used at device_attach() handler to initialize variables.

File: hconf.c [hunk 2]
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
...
+hconf_get_switch(struct hconf_softc *sc, int swtype, u_int *mask)
+{
+	uint8_t *fbuf;
+	int error;
+
+	if (sc->switch_rlen[swtype] <= 1)
+		return (ENXIO);
+
+	fbuf = malloc(sc->switch_rlen[swtype], M_TEMP, M_WAITOK | M_ZERO);
+	sx_xlock(&sc->lock);
```
Comment (wulf7) [inline hconf.c]: hid requests are serialized by transport backends now, so no need to protect them with extra locks.  hconf_set_input_mode_impl() takes sx_lock to serialize soft context updates.

File: hconf.c [hunk 3]
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
...
+		    sc->switch_rlen[swtype] - 1, &sc->switch_loc[swtype]);
+	}
+
+	sx_unlock(&sc->lock);
+	free(fbuf, M_TEMP);
+	return (error);
+}
+
+static int
+hconf_set_switch(struct hconf_softc *sc, int swtype, u_int mask)
```
Comment (wulf7) [inline hconf.c]: hconf_set_switch() should be called from device_resume handler too as suspend/resume cycle often resets touchpad settings.

**Review by wulf7 — COMMENTED** [review]
Looks mostly good. I left minor notes pinned to code lines.

**Comment by avg-I** [comment]
Thank you very much for the suggestions!
Time permitting, I will implement all of them and submit a pull request.

I am also thinking about how to integrate this feature into desktop environments.
I took KDE / plasma5 as an example. There, a touchpad can be turned off either via a global shortcut or via an applet. It seems that when it's done KDE sets a property "atom" of an input driver and that's up to the driver what to do next. In the case xf86-input-synaptics it seems that the driver simply starts ignoring input events without actually doing anything with the kernel driver or the hardware. So, I think that it would be nice to plug the switch feature into that. But doing it via an OS-specific sysctl does not seem to be nice. I wonder if there is a more generic interface (like a set of ioctls) for setting HID device features. I tried looking in Linux code but could not find anything of interest.

**Comment by wulf7** [comment]
> I will implement all of them and submit a pull request.

Great! Thank you!

> I wonder if there is a more generic interface (like a set of ioctls) for setting HID device features.

I do not know.
Hidraw/uhid can be used but they are too low level.
Unimplemented evdev's EVIOCSMASK ioctl looks promising, but it is per-client, so it requires support from all intermediate libraries like libevdev and libinput to work.
We can invent new input event codes also, but it requires tight coordination with upstream.

**Review by wulf7 — APPROVED** [review]
Looks good. Thank you.

**Pull request merged**